app/services/polymorphic_term_repository.rb

```diff
@@ -0,0 +1,36 @@
+class PolymorphicTermRepository < Struct.new(:id)
+  class << self
+    alias_method :orig_new, :new
+    def new(*args)
+      orig_new(*args).build
+    end
+
+    def find(*args)
+      orig_new(*args).find
+    end
+  end
+
+  def build
+    repository.new(id)
+  end
+
+  def find
+    repository.find(id)
+  end
+
+  private
+
+  def repository
+    return vocabulary_repository unless id.include?("/")
```

Comment:
Oh, or maybe not.. TermUri doesn't really do quite the same kind of operation, it's all about the ending of the id....  Hmm.